File: src/Core.Go/zset_merkle/zset_merkle.go
package zset_merkle

import (
	"encoding/binary"
	"fmt"
	"math/bits"
	"sort"

	"github.com/zeebo/xxh3"
)

type MerkleHash struct {
	Hi, Lo uint64
}

func (m MerkleHash) ToHex() string {
	return fmt.Sprintf("%016x%016x", m.Hi, m.Lo)
}

func OfBytes(b []byte) MerkleHash {
	h := xxh3.Hash128(b)
	return MerkleHash{
		Hi: bits.ReverseBytes64(h.Lo),
		Lo: bits.ReverseBytes64(h.Hi),
	}
}

func Combine(a, b MerkleHash) MerkleHash {
	var buf [32]byte
	binary.LittleEndian.PutUint64(buf[0:8], a.Hi)
	binary.LittleEndian.PutUint64(buf[8:16], a.Lo)
	binary.LittleEndian.PutUint64(buf[16:24], b.Hi)
	binary.LittleEndian.PutUint64(buf[24:32], b.Lo)
	return OfBytes(buf[:])
}

func leafBytes(keyBytes []byte, weight int64) []byte {
	buf := make([]byte, 4+len(keyBytes)+8)
	binary.LittleEndian.PutUint32(buf[0:4], uint32(len(keyBytes)))
	copy(buf[4:4+len(keyBytes)], keyBytes)
	binary.LittleEndian.PutUint64(buf[4+len(keyBytes):4+len(keyBytes)+8], uint64(weight))
	return buf
}

func fold(level []MerkleHash) MerkleHash {
	n := len(level)
	if n == 0 {
		return OfBytes([]byte{})
	}
	if n == 1 {
		return level[0]
	}
	parents := make([]MerkleHash, (n+1)/2)
	for i := 0; i < len(parents); i++ {
		a := level[2*i]
		var b MerkleHash
		if 2*i+1 < n {
			b = level[2*i+1]
		} else {
			b = a
		}
		parents[i] = Combine(a, b)
	}
	return fold(parents)
}

type Entry struct {
	Key    string
	Weight int64
}

func Root(entries []Entry) MerkleHash {
	counts := make(map[string]int64)
	for _, e := range entries {
		counts[e.Key] += e.Weight
	}

	var keys []string
	for k, w := range counts {
		if w != 0 {
			keys = append(keys, k)
		}
	}

	sort.Strings(keys)

	leaves := make([]MerkleHash, len(keys))
	for i, k := range keys {
		leaves[i] = OfBytes(leafBytes([]byte(k), counts[k]))
	}

	return fold(leaves)
}

File: src/Core.Python/src/zeta/zset_merkle.py
import struct

import xxhash


class MerkleHash:
    def __init__(self, hi: int, lo: int):
        self.hi = hi
        self.lo = lo

    @staticmethod
    def zero() -> MerkleHash:
        return MerkleHash(0, 0)

    @staticmethod
    def of_bytes(b: bytes) -> MerkleHash:
        digest = xxhash.xxh128(b).digest()
        lo = struct.unpack("<Q", digest[0:8])[0]
        hi = struct.unpack("<Q", digest[8:16])[0]
        return MerkleHash(hi, lo)

    @staticmethod
    def combine(a: MerkleHash, b: MerkleHash) -> MerkleHash:
        buf = bytearray(32)
        struct.pack_into("<Q", buf, 0, a.hi)
        struct.pack_into("<Q", buf, 8, a.lo)
        struct.pack_into("<Q", buf, 16, b.hi)
        struct.pack_into("<Q", buf, 24, b.lo)
        return MerkleHash.of_bytes(bytes(buf))

    def to_hex(self) -> str:
        return f"{self.hi:016x}{self.lo:016x}"


def leaf_bytes(key_bytes: bytes, weight: int) -> bytes:
    buf = bytearray(4 + len(key_bytes) + 8)
    struct.pack_into("<I", buf, 0, len(key_bytes))
    buf[4 : 4 + len(key_bytes)] = key_bytes
    struct.pack_into("<q", buf, 4 + len(key_bytes), weight)
    return bytes(buf)


def fold(level: list[MerkleHash]) -> MerkleHash:
    n = len(level)
    if n == 0:
        return MerkleHash.of_bytes(b"")
    if n == 1:
        return level[0]
    parents = []
    for i in range((n + 1) // 2):
        a = level[2 * i]
        b = level[2 * i + 1] if 2 * i + 1 < n else a
        parents.append(MerkleHash.combine(a, b))
    return fold(parents)


def root(entries: list[tuple[str, int]]) -> MerkleHash:
    # 1. Group by key and sum weights
    counts: dict[str, int] = {}
    for k, w in entries:
        counts[k] = counts.get(k, 0) + w

    # 2. Filter out zero net weights
    non_zero = [(k, w) for k, w in counts.items() if w != 0]

    # 3. Sort by key's encoded UTF-8 bytes (ordinal lexicographic)
    # Python sorts lists of tuples lexicographically; we sort key bytes
    encoded_entries = [(k.encode("utf-8"), w) for k, w in non_zero]
    encoded_entries.sort(key=lambda item: item[0])

    # 4. Generate leaf hashes
    leaves = [MerkleHash.of_bytes(leaf_bytes(kb, w)) for kb, w in encoded_entries]

    # 5. Fold bottom-up
    return fold(leaves)
